[DeviceMesh] Add extra check in flatten result cache lookup

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163358
* #161224
* #163213
* #163367
* __->__ #163288
* #163212

while refactoring DeviceMesh bookkeeping, we found that there is one corner case which we just don't check whether the dims to be flattened into is same as the dims which an existing flattened name maps to. So we need to add extra cases in the unit test and extra check logic in the code.



cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

fduwjj committed:
[DeviceMesh] Add extra check in flatten result cache lookup

[ghstack-poisoned]

wz337 commented:
LGTM! Thanks for adding extra check

fegin reviewed (approved):
Minor typo, which you can fix in another PR.

Skylion007 commented on torch/distributed/device_mesh.py:
```diff
@@ -185,7 +185,16 @@ def create_flatten_mesh(
+                ):
+                    return self.root_to_flatten_mapping[root_mesh][mesh_dim_name]
+                else:
+                    raise RuntimeError(
```
I think the latter two are more appropriate, right?

fduwjj committed:
Update on "[DeviceMesh] Add extra check in flatten result cache lookup"


while refactoring DeviceMesh bookkeeping, we found that there is one corner case which we just don't check whether the dims to be flattened into is same as the dims which an existing flattened name maps to. So we need to add extra cases in the unit test and extra check logic in the code.



cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Add extra check in flatten result cache lookup"


while refactoring DeviceMesh bookkeeping, we found that there is one corner case which we just don't check whether the dims to be flattened into is same as the dims which an existing flattened name maps to. So we need to add extra cases in the unit test and extra check logic in the code.



cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
